### ISSUE TITLE
Grad scale is small in Zipformer

### ISSUE THREAD

**Issue by TungyuYoung** [k2-fsa/icefall#1946]
Hi!

I tried to train a pretrained model for KWS. I used the Gigaspeech recipe and the XL subset. The parameters of the model are set as:

`
      --decoder-dim 320 \
      --joiner-dim 320 \
      --num-encoder-layers 1,1,1,1,1,1 \
      --feedforward-dim 192,192,192,192,192,192 \
      --encoder-dim 128,128,128,128,128,128 \
      --encoder-unmasked-dim 128,128,128,128,128,128 \
`

I used 8 * 4090 24G and set the base lr as 0.02 and MAX DURATION as 800. But this problem still happend. I use torch 2.4+cu124.

Is there any way to save this problem?

The training log is:

2025-05-20 02:38:47,195 WARNING [train_phone.py:1047] (7/8) Grad scale is small: 3.0517578125e-05
2025-05-20 02:38:47,195 INFO [train_phone.py:1056] (7/8) Epoch 1, batch 14300, loss[loss=0.6675, simple_loss=0.6511, pruned_loss=0.2501, ctc_loss=0.4594, over 19153.00 frames. ], tot_loss[loss=0.6085, simple_loss=0.5909, pruned_loss=0.2276, ctc_loss=0.4316, over 37589436.30 frames. ], batch size: 130, lr: 2.73e-02, grad_scale: 6.103515625e-05
2025-05-20 02:39:06,895 INFO [scaling.py:215] (7/8) ScheduledFloat: name=encoder.encoders.0.layers.0.bypass_mid.scale_min, batch_count=152746.66666666666, ans=0.2
2025-05-20 02:39:16,147 INFO [scaling.py:215] (7/8) ScheduledFloat: name=encoder.encoders.4.encoder.layers.0.bypass.scale_min, batch_count=152853.33333333334, ans=0.2
2025-05-20 02:39:16,183 INFO [scaling.py:1025] (7/8) Whitening: name=encoder.encoders.5.encoder.layers.0.conv_module2.whiten, num_groups=1, num_channels=128, metric=3.69 vs. limit=15.0
2025-05-20 02:39:27,036 WARNING [optim.py:604] (7/8) Scaling gradients by 0.43019792437553406, model_norm_threshold=1181.589599609375
2025-05-20 02:39:27,089 WARNING [optim.py:726] (7/8) Parameter dominating tot_sumsq module.joiner.output_linear.weight with proportion 0.36, where dominant_sumsq=(grad_sumsq*orig_rms_sq)=2.753e+06, grad_sumsq=1.151e+08, orig_rms_sq=2.391e-02
2025-05-20 02:39:27,096 WARNING [optim.py:665] (7/8) Parameters with most larger-than-usual grads, with ratios, are: [(15.611163139343262, 'module.encoder.encoders.5.encoder.layers.0.norm.log_scale', []), (8.762660026550293, 'module.joiner.output_linear.weight', [0, 226]), (6.9120330810546875, 'module.encoder.downsample_output.bias', [1]), (5.065295219421387, 'module.encoder.encoders.5.encoder.layers.0.self_attn1.in_proj.bias', [33]), (4.653099536895752, 'module.encoder.encoders.4.encoder.layers.0.self_attn2.in_proj.bias', [39]), (4.28786039352417, 'module.encoder_embed.out_norm.bias', [37]), (4.143716812133789, 'module.encoder.encoders.5.encoder.layers.0.self_attn2.in_proj.bias', [14]), (3.827667474746704, 'module.encoder_embed.conv.0.weight', [2, 0, 0, 2]), (3.7848217487335205, 'module.encoder.encoders.4.downsample.bias', [0]), (3.507585048675537, 'module.encoder.encoders.2.encoder.layers.0.self_attn1.in_proj.bias', [15])]
2025-05-20 02:39:36,241 INFO [train_phone.py:1056] (7/8) Epoch 1, batch 14350, loss[loss=0.7004, simple_loss=0.6748, pruned_loss=0.2617, ctc_loss=0.5065, over 19244.00 frames. ], tot_loss[loss=0.6107, simple_loss=0.5927, pruned_loss=0.2286, ctc_loss=0.4335, over 37570857.65 frames. ], batch size: 140, lr: 2.72e-02, grad_scale: 3.0517578125e-05
2025-05-20 02:39:40,024 INFO [scaling.py:215] (7/8) ScheduledFloat: name=encoder.encoders.2.encoder.layers.0.ff3_skip_rate, batch_count=153066.66666666666, ans=0.0
2025-05-20 02:39:56,239 WARNING [optim.py:604] (7/8) Scaling gradients by 0.4739697575569153, model_norm_threshold=1181.589599609375
2025-05-20 02:39:56,292 WARNING [optim.py:726] (7/8) Parameter dominating tot_sumsq module.encoder.encoders.5.encoder.layers.0.self_attn_weights.in_proj.bias with proportion 0.63, where dominant_sumsq=(grad_sumsq*orig_rms_sq)=3.945e+06, grad_sumsq=1.243e+07, orig_rms_sq=3.173e-01
2025-05-20 02:39:56,300 WARNING [optim.py:665] (7/8) Parameters with most larger-than-usual grads, with ratios, are: [(19.62128257751465, 'module.encoder.encoders.5.encoder.layers.0.self_attn_weights.linear_pos.weight', [0, 47]), (11.576221466064453, 'module.encoder.encoders.4.encoder.layers.0.norm.log_scale', []), (4.264790058135986, 'module.encoder.encoders.5.downsample.bias', [0]), (3.974349021911621, 'module.encoder.encoders.4.encoder.layers.0.self_attn2.in_proj.bias', [44]), (2.9069502353668213, 'module.encoder.encoders.4.encoder.layers.0.self_attn1.in_proj.bias', [2]), (2.8130431175231934, 'module.encoder.encoders.2.downsample.bias', [0]), (2.733820915222168, 'module.encoder.encoders.2.encoder.layers.0.norm.log_scale', []), (2.7192320823669434, 'module.encoder.encoders.4.encoder.layers.0.self_attn2.in_proj.weight', [8, 29]), (2.603834629058838, 'module.encoder.encoders.1.encoder.layers.0.norm.log_scale', []), (2.3517298698425293, 'module.encoder.encoders.4.encoder.layers.0.conv_module2.depthwise_conv.chunkwise_conv_scale', [0, 49, 3])]
2025-05-20 02:40:11,883 WARNING [optim.py:588] (7/8) Clipping_scale=2.0, grad-norm quartiles 3.701e+02 4.680e+02 6.219e+02 9.996e+02 2.747e+03, threshold=1.244e+03, percent-clipped=18.0
2025-05-20 02:40:22,372 WARNING [train_phone.py:1047] (7/8) Grad scale is small: 1.52587890625e-05
2025-05-20 02:40:22,373 INFO [train_phone.py:1056] (7/8) Epoch 1, batch 14400, loss[loss=0.7653, simple_loss=0.6959, pruned_loss=0.3063, ctc_loss=0.5553, over 19276.00 frames. ], tot_loss[loss=0.6143, simple_loss=0.5951, pruned_loss=0.2303, ctc_loss=0.4366, over 37569811.71 frames. ], batch size: 158, lr: 2.72e-02, grad_scale: 3.0517578125e-05
2025-05-20 02:40:29,360 INFO [scaling.py:215] (7/8) ScheduledFloat: name=encoder.encoders.4.encoder.layers.0.self_attn_weights.pos_emb_skip_rate, batch_count=153706.66666666666, ans=0.0
2025-05-20 02:40:43,161 WARNING [optim.py:604] (7/8) Scaling gradients by 0.3832685649394989, model_norm_threshold=1243.713134765625

**Comment by TungyuYoung** [comment]
By the way, I use CTC-loss

**Comment by TungyuYoung** [comment]
@pkufool Thanks! I will verify it these feel days. I will comment you after all things finished.
